### PR TITLE
Get set particle positions

### PR DESCRIPTION
**Pull request by luca-heltai**
Get and set particle positions from various sources.

Depends on #9074

### CODE REVIEW

**Comment by bangerth** [comment]
Same question here as in #9076.

**Comment by blaisb** [comment]
> Same question here as in #9076.

Once #9073 is merged I will take care of this pull request. I think it is better if I go through it once before you start taking your time to review it. I will take care of moving the implementation to the .cc file and etc.

**Comment by bangerth** [comment]
OK. I'm slightly confused by which PRs are current and ready and which are not. So if you'd like me to look at any one in particular, please note that in a comment there.

**Comment by blaisb** [comment]
> OK. I'm slightly confused by which PRs are current and ready and which are not. So if you'd like me to look at any one in particular, please note that in a comment there.

I think Luca and I are still confused that about also. We made 9 pull request on the same day after two weeks of relatively intense work and some of the routines have evolved significantly since then (for instance the global insertion of particles). Let me ping you when pull requests are in a review-ready shapes. It won't be much work on my part, but it will save time and prevent work duplication.

**Comment by luca-heltai** [comment]
@blaisb, I'm making some changes to this branch. Are you also working here?

I'm removing the utilities.h file, and inserting all the get/set functions directly in the ParticleHandler class. This should be ready today or tomorrow at the latest.

**Comment by blaisb** [comment]
> @blaisb, I'm making some changes to this branch. Are you also working here?
> 
> I'm removing the utilities.h file, and inserting all the get/set functions directly in the ParticleHandler class. This should be ready today or tomorrow at the latest.

@luca-heltai I only had rebased to remove the merge conflict. I was actually going to work on this today or tomorrow. Go ahead and make the changes and I will gladly look at them after you have committed them. You seem to have a more thorough plan in mind than me :)!

**Comment by luca-heltai** [comment]
@bangerth , @blaisb , this is now ready for a review. As pointed out, I put the functionality right into ParticleHandler, and adapted the interface to the latest status of the previous commits.

File: include/deal.II/particles/particle_handler.h [hunk 1]
```diff
@@ -619,6 +750,58 @@ namespace Particles
...
+  void
+  ParticleHandler<dim, spacedim>::get_particle_positions(
+    VectorType &output_vector,
+    const bool  add_to_output_vector)
+  {
+    AssertDimension(output_vector.size(),
+                    get_next_free_particle_index() * spacedim);
+    for (const auto &p : *this)
+      {
+        auto &     point = p.get_location();
```
Comment (bangerth) [inline include/deal.II/particles/particle_handler.h]: It's a point, let's just copy it.
```suggestion
        auto point = p.get_location();
```

File: include/deal.II/particles/particle_handler.h [hunk 2]
```diff
@@ -619,6 +750,58 @@ namespace Particles
...
+          for (unsigned int i = 0; i < spacedim; ++i)
+            output_vector[id * spacedim + i] += point[i];
+        else
+          for (unsigned int i = 0; i < spacedim; ++i)
+            output_vector[id * spacedim + i] = point[i];
+      }
+    if (add_to_output_vector)
+      output_vector.compress(VectorOperation::add);
+    else
+      output_vector.compress(VectorOperation::insert);
```
Comment (bangerth) [inline include/deal.II/particles/particle_handler.h]: Ah, so it's a very specific kind of vector type that supports the `compress()` operation.

File: source/particles/particle_handler.cc [hunk 3]
```diff
@@ -719,6 +719,77 @@ namespace Particles
...
+    AssertDimension(function.n_components, spacedim);
+
+    Vector<double> new_position(spacedim);
+    for (auto &particle : *this)
+      {
+        Point<spacedim> particle_location = particle.get_location();
+        function.vector_value(particle_location, new_position);
+        if (displace_particles)
+          for (unsigned int d = 0; d < spacedim; ++d)
+            particle_location[d] = particle_location[d] + new_position[d];
```
Comment (bangerth) [inline source/particles/particle_handler.cc]:
```suggestion
            particle_location[d] += new_position[d];
```

File: include/deal.II/particles/particle_handler.h [hunk 4]
```diff
@@ -285,6 +286,136 @@ namespace Particles
...
+     * @p input_vector. By default, the particles are displaced of the
+     * amount contained in the @p input_vector.
+     *
+     * After setting the new position, this function calls internally the method
+     * sort_particles_into_subdomains_and_cells(). You should
+     * make sure you satisfy the requirements of that function.
+     *
+     * @param[in] input_vector A parallel distributed vector containing
+     * the displacement to apply to each particle, or their new absolute
+     * position.
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/particles/particle_handler.h]: It says so in the documentation few lines above:

~~~

     * Set the position of the particles by using the values contained in the
     * vector @p input_vector.
     *
     * The vector @p input_vector should have read access to the indices
     * created by extracting the locally relevant ids with
     * locally_relevant_ids(), and taking its tensor
     * product with the index set representing the range `[0, spacedim)`, i.e.:
     * @code
     * IndexSet ids = particle_handler.locally_relevant_ids().
     *  tensor_product(complete_index_set(spacedim));
     * @endcode
~~~

File: include/deal.II/particles/particle_handler.h [hunk 5]
```diff
@@ -285,6 +286,136 @@ namespace Particles
...
+     *
+     * @param[in] displace_particles Control if the @p input_vector should
+     * be interpreted as a displacement vector, or a vector of absolute
+     * positions.
+     *
+     * @authors Luca Heltai, Bruno Blais, 2019.
+     */
+    template <class VectorType>
+    void
+    set_particle_positions(const VectorType &input_vector,
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/particles/particle_handler.h]: Sure. But that does not allow linear algebra operations to be performed with PETSc, Trilinos, etc.

I'll let that to anyone else needing that particular feature. We use this with velocity vectors, and use ODEs and DAEs defined on `VectorType` coming from either deal.II/Trilinos/or PETSc.

File: include/deal.II/particles/particle_handler.h (hunk 2 above)
Comment (luca-heltai) [reply include/deal.II/particles/particle_handler.h]: All of our vector types support the compress() operation.

File: include/deal.II/particles/particle_handler.h [hunk 6]
```diff
@@ -285,6 +286,136 @@ namespace Particles
...
+     * particles.
+     *
+     * @param[in] add_to_output_vector Control if the function should set the
+     * entries of the @p output_vector or if should add to them.
+     *
+     * @author Luca Heltai, Bruno Blais, 2019.
+     */
+    template <class VectorType>
+    void
+    get_particle_positions(VectorType &output_vector,
```
[the remark this answers is not in the document]
Comment (luca-heltai) [reply include/deal.II/particles/particle_handler.h]: It accepts all the LAC vectors currently implemented in deal.II. Should we specify this?

File: include/deal.II/particles/particle_handler.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/particles/particle_handler.h]: I meant to say: It seems like this is a very generic data type, but you're thinking of this specifically as the *deal.II distributed vectors*, including PETSc and Trilinos wrappers that are partitioned in such a way that the entries we need above are locally owned. (Do they need to have particular ghost elements? I don't think so.)

Would you humor me and put that into a `@tparam` description for `VectorType`?

File: include/deal.II/particles/particle_handler.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/particles/particle_handler.h]: Yes, if you don't mind -- just copy the `@tparam` section I mentioned above.

File: include/deal.II/particles/particle_handler.h [hunk 7]
```diff
@@ -285,6 +286,146 @@ namespace Particles
...
+                           const bool  add_to_output_vector = false);
+
+    /**
+     * Gather the position of the particles within the particle handler in
+     * a vector of points. The order of the points is the same on would obtain
+     * by iterating over all (local) particles, and querying their locations.
+     *
+     * @param [in,out] positions A vector preallocated at size
+     * (particle_handler.n_locally_owned_articles) and whose points will become
```
Comment (bangerth) [inline include/deal.II/particles/particle_handler.h]:
```suggestion
     * `particle_handler.n_locally_owned_articles()` and whose points will become
```

File: include/deal.II/particles/particle_handler.h [hunk 8]
```diff
@@ -619,6 +760,58 @@ namespace Particles
...
+    const VectorType &input_vector,
+    const bool        displace_particles)
+  {
+    AssertDimension(input_vector.size(),
+                    get_next_free_particle_index() * spacedim);
+    for (auto &p : *this)
+      {
+        auto       new_point(displace_particles ? p.get_location() :
+                                            Point<spacedim>());
+        const auto id = p.get_id();
```
Comment (bangerth) [inline include/deal.II/particles/particle_handler.h]: What a bizarre indentation...

**Pull request merged**